### ISSUE TITLE
SolidStart file routing

### ISSUE THREAD

**Issue by haodeon** [fable-compiler/Fable.Solid#21]
I'm exploring SolidStart and I thought I would also try Fable at the same time 😄.

I had some difficulty with Solid's file-based routing. The docs state each route "must default export a component" and it appears the Solid compiler is specifically looking for `export default function Home() {`.

The trouble I had was I couldn't figure out a way to `export default function ...` without an emit. With `exportDefault`, the generated JS would end up being `export default (Home)`, with Home being defined earlier as a const. Using emit though I couldn't figure out how to parameterise the function name and make a generic helper.

What I ended up doing was making the component function inline so `let inline Home () =` and using `exportDefault`. This generates a default arrow function `export default (() =>` which works with SolidStart.

Is there a better way to do this?

**Comment by MangelMaxime** [comment]
I am not familiar with Solid so I am not sure which exact JavaScript it needs or if there is some telerance.

But if you want a bit of type safety regarding the name of the function you could do something like that:

```fs
open Fable.Core
open Fable.Core.JsInterop


let inline Home () = 
    "Hello"

let inline exportDefaultSolidJs (name : string) (compFunc : obj)=
    emitJsStatement (name, compFunc) $$"""
export default function $0 {
    return $1
}
    """

exportDefaultSolidJs (nameof Home) Home
```

It generates

```js
export default function "Home" {
    return (() => "Hello")
}
    ;
```

It has the benefit when if you change `Home` name you will get a compiler warning or have it done automatically if you rename the symbol.

**Comment by haodeon** [comment]
Thanks for the suggestion, I really appreciate the assistance.

Unfortunately, the generated output runs into an issue I'd already encountered while attempting to parameterise the function name.

`<stdin>:3:24: ERROR: Expected "(" but found "\"Home\""`

If I change the export function to

```fs
let inline exportDefaultSolidJs (name: string) (compFunc: obj) =
    emitJsStatement
        (name, compFunc)
        $$"""
export default function $0 () {
    return $1
}
    """
```

It still fails with same error.

The closest I have got is with a variant of something I have tried previously plus your suggestion

```fs
[<Global>]
let Home () = jsNative

let inline home () =
    Html.main [
        Html.children [
            JSX.create Title [ Html.children [ Html.text "About" ] ]
            Html.h1 [ Html.children [ Html.text "About" ] ]
        ]
    ]

let inline exportDefaultSolidJs (name: obj) (func: obj) =
    emitJsStatement
        (name, func)
        $$"""
export default function $0 {
    return $1
}"""

exportDefaultSolidJs Home home
```

This generates

```js
export default function (Home) {
    return (() => <main>
    <Title>
        About
    </Title>
    <h1>
        About
    </h1>
</main>)
};
```

Which initially appears to work when I click the "About" link but refreshing the page or trying to directly access the URL causes the client to error with

`Hydration Mismatch. Unable to find DOM nodes for hydration key: 0000000001000002000001`

I believe the cause is the return being an anonymous function instead of a component. I haven't figured out how to inline values.

**Comment by MangelMaxime** [comment]
I yes I didn't see that it generates `"Home"` and not `Home`.

Hum, I don't have a better idea right now.

The only one, I have is to perhaps look into using Fable.CompilerPlugin system which allows you to change how Fable emit the code. It allows you to hook into the compiler itself, so you can "in theory do anything". However, I am not too familiar with this part of Fable.

You can look at how [Feliz.CompilerPlugins](https://github.com/Zaid-Ajaj/Feliz/tree/master/Feliz.CompilerPlugins) is implemented to have a feeling of the API.

By using the plugin system you could in theory:

1. Decorate a function with an attribute
2. Capture said functions
3. Rewrite the emitted surrounding of the function, to match with what Sold expects

**Comment by MangelMaxime** [comment]
Actually, I have another idea.

The generated code is returning a function that needs to be called.

Does making the function an IIFE (Immediately Invoked Function Expression) works?

```js
export default function (Home) {
    return (() => <main>
    <Title>
        About
    </Title>
    <h1>
        About
    </h1>
</main>)()
};
```

If yes then you could use:

```fs
let inline exportDefaultSolidJs (name: obj) (func: obj) =
    emitJsStatement
        (name, func)
        $$"""
export default function $0 {
    return $1()
}"""
```

Note the extra `()` on the return line.

The only question then is does all components takes zero arguments or do they need some sometimes?

**Comment by haodeon** [comment]
I've just tested adding the extra `()` and yes it works.

Thank you for helping figure this one out. It's great to have finally nailed this.
Reviewing the solution though, I feel the use of `Global` attribute and `emitJsStatement` is far more complex to understand than an inline to `exportDefault`. I'll use this though when I need to.

To answer your question, file route components do sometimes need arguments. `(props: RouteSectionProps)` is the first one to come to mind.

Closing this issue as I feel it's been answered. It's an odd intersection between Fable transpiling to JS and how Solid compiles components.
